### ISSUE TITLE
How to process a video frame to be ready for augmented reality

### ISSUE THREAD

**Issue by wangyong01** [hjwdzh/FrameNet#2]
Dear authors, appreciate for your amazing work.

I tried the AR demo and find that orient-X_pred and Y are provided for each image (or video frame), and wondering the pipeline for preprocessing a image to be ready for the ar demo.

Thanks.

**Comment by hjwdzh** [comment]
orient_X/Y_pred can be directly copied from the network prediction. You could take a look at our training code.

**Comment by wangyong01** [comment]
Sorry for bather you again on this issue.
I have tried to run the model to get orient-X_pred and Y. But I guess there maybe some following smoothing steps, as I used the demo 0000-color for testing and get results as yours.

The first one is using output[:, 0:3, :, :] with checkpoints on DORN(channel=5, output_channel=13)
The second one is orient-X_pred from your demo.
Best, thanks for your time.
![image](https://user-images.githubusercontent.com/39072599/81473690-c6d3a280-9232-11ea-8764-872520a712ca.png)
![image](https://user-images.githubusercontent.com/39072599/81473694-ce934700-9232-11ea-9f26-d1547fe19fe2.png)

**Comment by hjwdzh** [comment]
Not sure what you are asking. All demo test images are direct outputs from the network. There is no additional smoothing step.

**Comment by wangyong01** [comment]
Thanks for the quick reply and sorry for my vague question.
I tried as your direction for outputting dir-x and dir-y images, but the results are different, and the attachTexture not work on these images. Here following is the code.

```
def read_images():
    intrinsics = [577.591, 318.905, 578.73, 242.684]
    xx, yy = np.meshgrid(np.array([i for i in range(640)]), np.array([i for i in range(480)]))
    mesh_x = misc.imresize((xx - intrinsics[1]) / intrinsics[0], (240, 320), 'nearest', mode='F')
    mesh_y = misc.imresize((yy - intrinsics[3]) / intrinsics[2], (240, 320), 'nearest', mode='F')
    start_name = 'results/0000'
    color_name = start_name + '-color.png'
    color_img = misc.imresize(sio.imread(color_name), (240, 320, 3), 'nearest')
    to_tensor = transforms.ToTensor()
    color_tensor = to_tensor(color_img)
    input_tensor = torch.zeros((5, color_img.shape[0], color_img.shape[1]), dtype=torch.float32)
    input_tensor[0:3, :, :] = color_tensor
    input_tensor[3, :, :] = torch.Tensor(mesh_x)
    input_tensor[4, :, :] = torch.Tensor(mesh_y)
    return input_tensor.unsqueeze(0)
def Normalize(dir_x):
    dir_x_l = torch.sqrt(torch.sum(dir_x ** 2, dim=1) + 1e-7).view(dir_x.shape[0], 1, dir_x.shape[2], dir_x.shape[3])
    dir_x_l = torch.cat([dir_x_l, dir_x_l, dir_x_l], dim=1)
    return dir_x / dir_x_l
def main():
    model = DORN(channel=5, output_channel=13)
    model = model.cuda()
    model.eval()
    model.load_state_dict(torch.load('checkpoints/dorn.cpkt'))
    image = read_images()
    image = image.cuda()
    with torch.no_grad():
        output = model(image)
    save_image(Normalize(output[:, 4:7, :, :]), 'results/0000-orient-X_pred.png')
    save_image(Normalize(output[:, 7:10, :, :]), 'results/0000-orient-Y_pred.png')

```
Following are my orient-X_pred.png and orient-X_pred.png gets from the demo selected dir.
![image](https://user-images.githubusercontent.com/39072599/81491436-31cbba80-92c1-11ea-8670-acab6e7225ed.png)
![image](https://user-images.githubusercontent.com/39072599/81491467-84a57200-92c1-11ea-9a83-7be083026283.png)

**Comment by hjwdzh** [comment]
Did you first sanity check to confirm that your code generate the same results for the provided demo images?

I did not carefully look at your provided code, but seems that you did not normalize value to [0,1] before saving as images.

If you take a look at how I load images in AttachTexture.py, you will see that I load the input image and recover the direction using Color2Vec. You will need a inverse version of Color2Vec to save direction to images.

**Comment by wangyong01** [comment]
Thanks. Got it!

**Comment by lartpang** [comment]
@hjwdzh  @wangyong01  I use a similar code, but I can't get a good orient image. Can you provide some suggestions?

```python
import os

import numpy as np
import torch
from PIL import Image
from skimage import io as sio
from torchvision.transforms import transforms

from dorn import DORN


def imresize(data, size, interp, mode=None):
    return np.array(Image.fromarray(data, mode=mode).resize(size=size, resample=interp))


def read_images(color_name):
    intrinsics = [577.591, 318.905, 578.73, 242.684]
    xx, yy = np.meshgrid(
        np.array([i for i in range(640)]), np.array([i for i in range(480)])
    )
    mesh_x = imresize(
        (xx - intrinsics[1]) / intrinsics[0],
        (320, 240),
        Image.NEAREST,
        mode="F",
    )
    mesh_y = imresize(
        (yy - intrinsics[3]) / intrinsics[2],
        (320, 240),
        Image.NEAREST,
        mode="F",
    )

    color_img = sio.imread(color_name)
    ori_shape = color_img.shape
    color_img = imresize(color_img, (320, 240), Image.BILINEAR)
    to_tensor = transforms.ToTensor()
    color_tensor = to_tensor(color_img)
    input_tensor = torch.zeros(
        (5, color_img.shape[0], color_img.shape[1]), dtype=torch.float32
    )
    input_tensor[0:3, :, :] = color_tensor
    input_tensor[3, :, :] = torch.Tensor(mesh_x)
    input_tensor[4, :, :] = torch.Tensor(mesh_y)
    return input_tensor.unsqueeze(0), ori_shape


def Normalize(dir_x):
    dir_x_l = torch.sqrt(torch.sum(dir_x ** 2, dim=1) + 1e-7).view(
        dir_x.shape[0], 1, dir_x.shape[2], dir_x.shape[3]
    )
    dir_x_l = torch.cat([dir_x_l, dir_x_l, dir_x_l], dim=1)
    return dir_x / dir_x_l


def main():
    model = DORN(channel=5, output_channel=13)
    model.load_state_dict(torch.load("checkpoints/dorn.cpkt", map_location="cpu"))
    model.eval()
    print("model OK")

    data_name = "0000"
    data_root = "demo/selected/{}-color.png"
    image, ori_shape = read_images(color_name=data_root.format(data_name))
    with torch.no_grad():
        output = model(image)

    if not os.path.exists("results"):
        print("results", " does not exist.")
        os.makedirs("results")

    dir_x = Normalize(output[:, 4:7, :, :])
    dir_y = Normalize(output[:, 7:10, :, :])
    dir_x = Vec2Color(dir_x.squeeze().permute(1, 2, 0).numpy())
    dir_y = Vec2Color(dir_y.squeeze().permute(1, 2, 0).numpy())
    dir_x = imresize(dir_x, ori_shape[:2][::-1], Image.BILINEAR)
    dir_y = imresize(dir_y, ori_shape[:2][::-1], Image.BILINEAR)
    sio.imsave("results/{}-orient-X_pred.png".format(data_name), dir_x)
    sio.imsave("results/{}-orient-Y_pred.png".format(data_name), dir_y)


def Vec2Color(data):
    data = (data + 1) / 2
    return (data * 255).astype(np.uint8)


if __name__ == "__main__":
    main()
```

Here is my orient-X_pred:

![0000-orient-X_pred](https://user-images.githubusercontent.com/26847524/120077724-6bdf4880-c0de-11eb-9413-319777c1f69c.png)

and orient-Y_pred:

![0000-orient-Y_pred](https://user-images.githubusercontent.com/26847524/120077753-7bf72800-c0de-11eb-8722-e055644bb3ba.png)
